### PR TITLE
feat: activate unification for effect sets

### PR DESCRIPTION
**Pull request by mlutze**
Related to:
* #4050 

Depends on:
* #4145

### CODE REVIEW

**Comment by mlutze** [comment]
Breaks a ton of stuff in the standard library. I think we need to move to the NNF form to avoid the explosion of type sizes.

**Comment by magnus-madsen** [comment]
We can talk tomorrow about what to do. Probably using the BoolMin table is the way to go. (If sound).

**Comment by magnus-madsen** [comment]
Will you write a status on what works? and performance?

**Comment by mlutze** [comment]
Everything _seems_ to work as expected. Some extra casts are needed in some cases, but that's mostly when we need to cast for Boolean Unification, for example when we have a continuation that uses the identity.

I haven't added negative tests for the new expressions yet, but the positive tests work, plus all the rest of the tests and the standard library with some modifications.

As for performance, I'll see what my machine can do, but I'm pretty skeptical of the results it produces.

**Comment by magnus-madsen** [comment]
Do you have `without` and does it give interesting results?

**Comment by magnus-madsen** [comment]
Should we merge under a flag? And are there things to refactor before? 

Then after we look for commonalities?

**Comment by magnus-madsen** [comment]
!!!jar

**Comment by mlutze** [comment]
> Do you have `without` and does it give interesting results?

`without` indeed exists, though I have not done much (maybe any) testing with it, since my focus has been "don't break everything else" :)

**Comment by mlutze** [comment]
> I am guessing there is no VSCode hovering of effects yet. Otherwise this would be interesting:
> 
> ```
> pub def foo(): Unit = 
>     let f = ???;
>     let g = ???;
>     let c = f >> g;
>     c() without MyEff
> ```

Right. Effects are erased after typing but it shouldn't be difficult to add that back.

**Comment by magnus-madsen** [comment]
Do we add it in this PR or the next?

**Comment by magnus-madsen** [comment]
So after benchmarks, we can merge this.

**Comment by mlutze** [comment]
> So after benchmarks, we can merge this.

Benchmarks showed no significant difference.

```text
matt@matt-laptop:~/Documents/flix-home/flix/personal/benchmarks$ ./benchmark feature
====================== Flix Compiler Phases ======================              

Mean runtime per phase (low to high):
  EntryPoint                       0.0 ms (00.0%)
  Reader                           2.8 ms (00.0%)
  Deriver                          3.7 ms (00.0%)
  Tailrec                         11.7 ms (00.1%)
  Safety                          24.7 ms (00.3%)
  VarNumbering                    26.5 ms (00.3%)
  Eraser                          32.1 ms (00.4%)
  Regions                         38.2 ms (00.4%)
  Simplifier                      40.8 ms (00.5%)
  LambdaLift                      42.2 ms (00.5%)
  Stratifier                      42.2 ms (00.5%)
  PatternExhaustiveness           50.4 ms (00.6%)
  Kinder                          86.5 ms (01.0%)
  Lowering                       101.3 ms (01.1%)
  TreeShaker                     105.4 ms (01.2%)
  ClosureConv                    143.6 ms (01.6%)
  Weeder                         147.6 ms (01.6%)
  Redundancy                     162.7 ms (01.8%)
  Instances                      184.2 ms (02.1%)
  Resolver                       214.7 ms (02.4%)
  Finalize                       219.1 ms (02.4%)
  Namer                          309.3 ms (03.4%)
  Parser                         496.3 ms (05.5%)
  Optimizer                      602.6 ms (06.7%)
  Monomorph                      1061.2 ms (11.8%)
  JvmBackend                     1880.3 ms (20.9%)
  Typer                          2952.6 ms (32.9%)

Finished 10 iterations on 60,762 lines of code in 71 seconds.
matt@matt-laptop:~/Documents/flix-home/flix/personal/benchmarks$ ./benchmark master
====================== Flix Compiler Phases ======================              

Mean runtime per phase (low to high):
  EntryPoint                       0.0 ms (00.0%)
  Reader                           3.3 ms (00.0%)
  Deriver                          3.7 ms (00.0%)
  Tailrec                         11.2 ms (00.1%)
  Safety                          16.2 ms (00.2%)
  VarNumbering                    26.7 ms (00.3%)
  Stratifier                      31.5 ms (00.3%)
  Regions                         39.2 ms (00.4%)
  LambdaLift                      39.8 ms (00.4%)
  Eraser                          48.2 ms (00.5%)
  Simplifier                      49.7 ms (00.5%)
  PatternExhaustiveness           66.3 ms (00.7%)
  Kinder                          93.0 ms (01.0%)
  TreeShaker                     103.1 ms (01.1%)
  Lowering                       109.1 ms (01.1%)
  ClosureConv                    143.1 ms (01.5%)
  Weeder                         173.1 ms (01.8%)
  Redundancy                     183.4 ms (01.9%)
  Instances                      199.2 ms (02.1%)
  Finalize                       209.8 ms (02.2%)
  Resolver                       266.3 ms (02.7%)
  Namer                          360.8 ms (03.7%)
  Optimizer                      583.3 ms (06.0%)
  Parser                         602.4 ms (06.2%)
  Monomorph                      1092.4 ms (11.3%)
  JvmBackend                     1985.0 ms (20.5%)
  Typer                          3243.5 ms (33.5%)

Finished 10 iterations on 60,762 lines of code in 76 seconds.
```

File: main/src/library/Nel.flix [hunk 1]
```diff
@@ -409,14 +409,14 @@ namespace Nel {
     /// Applies `g` to the initial element of `l` and combines it
     /// with the remainder of `l` using `f` going from right to left.
     ///
-    pub def reduceRightTo(f: (a, b) -> b & ef1, g: a -> b & ef2, l: Nel[a]): b & (ef1 and ef2) =
+    pub def reduceRightTo(f: (a, b) -> b & ef1 \ ef1_, g: a -> b & ef2 \ ef2_, l: Nel[a]): b & (ef1 and ef2) \ (ef1_ + ef2_) =
```
Comment (magnus-madsen) [inline main/src/library/Nel.flix:412]: Whats with the signature?

Comment (magnus-madsen) [reply main/src/library/Nel.flix:412]: And is a TODO needed?

Comment (mlutze) [reply main/src/library/Nel.flix:412]: The continuation needs to use the effect part of the type, but that is not available when the type is implicutly split. so we have to add the explicit purity and effect separately in the signature.

**Comment by magnus-madsen** [comment]
I guess we should rapidly add a flag and set it to off by default.

**Pull request merged**